### PR TITLE
`Learning paths`: Allow instructors to include prerequisites at the start

### PR DESCRIPTION
**Pull request by JohannesStoehr**
🚨 🚨 Delete the prerequisites created with this PR after testing 🚨 🚨 

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.

### Motivation and Context

#8765 added the creation and edit options for prerequisites. They should now get more integrated into the rest of Artemis like linking to learning objects and scheduling those in the learning path.

### Description

This PR adds the following things:
- You can now link prerequisites when editing exercises/lecture units
- You can now add relations between prerequisites themselves as well as to competencies
- Learning Paths now include prerequisites and their learning objects
- Started some refactoring from Competencies to CourseCompetencies (common super class of Competencies and Prerequisites)

What this PR does NOT cover:
- Linking lecture units when editing prerequisites
- UI changes apart from also showing prerequisites in relevant places (any visual differentiation to competencies will follow in FollowUps)
- Fixing the guided mode of lectures where it is currently not possible to link competencies
- Refactor all code from Competencies to CourseCompetencies

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Course with activated learning paths

1. Log in to Artemis
2. Navigate to Course Administration
3. Create one or more prerequisites and one or more competencies
4. Create relations between prerequisites and competencies
5. Create lecture units and exercises and link them to the prerequisites
6. Participate as a student in the learning path
   - If you want to check your mastery progress, this is currently only visible in the competency relation graph in the drop down for the learning path navigation ui
8. Delete a prerequisite as instructor
9. 🚨 🚨 Delete all other prerequisites that you created during the test of this PR 🚨 🚨 

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| competency.service.ts | 90.78% | ✅ |
| course-competency.service.ts | 100% | ✅ |
| create-competency.component.ts | 97.67% | ✅ |
| competency-selection.component.ts | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| LearningObjectService.java | 89% | ✅ |
| CompetencyProgressService.java | 94% | ✅ |
| CompetencyRelationService.java | 100% | ✅ |
| CompetencyService.java | 92% | ✅ |
| CourseCompetencyService.java | 86% | ✅ |
| PrerequisiteService.java | 77% | ✅ |
| LearningPathNavigationService.java | 92% | ✅ |
| LearningPathNgxService.java | 97% | ✅ |
| LearningPathRecommendationService.java | 98% | ✅ |
| LearningPathService.java | 97% | ✅ |
| CourseResource.java | 89% | ✅ |
| LearningPathResource.java | 100% | ✅ |
| CourseCompetencyResource.java | 100% | ✅ |
| PrerequisiteResource.java | 96% | ✅ |

## Summary by CodeRabbit

- **New Features**
  - Enhanced course information retrieval by including prerequisites in various services and models.

- **Refactor**
  - Renamed `Competency` to `CourseCompetency` across multiple classes and services for improved clarity and consistency.

- **Bug Fixes**
  - Adjusted method signatures and repository references to accommodate the new `CourseCompetency` entity.

- **Tests**
  - Updated test cases to align with the new `CourseCompetency` entity and added prerequisite handling in course retrieval tests.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:762]: We have the `@NotNull` annotation for exercises here

File: src/main/java/de/tum/in/www1/artemis/repository/CourseCompetencyRepository.java
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/repository/CourseCompetencyRepository.java:10]: It is used in `findAllExerciseInfoByCompetencyId`

File: src/main/java/de/tum/in/www1/artemis/web/rest/competency/CompetencyResource.java
Comment (pzdr7) [inline src/main/java/de/tum/in/www1/artemis/web/rest/competency/CompetencyResource.java:567]: I think you could update the name as well (this method no longer checks the authorization, but if the competency matches the course)

**Review by pzdr7 — APPROVED** [review]
Code looks good. One optional comment

**Review by pzdr7 — APPROVED** [review]

**Review by iyannsch — CHANGES_REQUESTED** [review]
The Competency Relations graph currently allows for circular dependencies between both competency<->competency, but also competency<->prerequisite and prerequisite<->prerequisite. For all combinations, it works with `Relates` but sometimes even with `Matches`.

**Review by egekurt123 — CHANGES_REQUESTED** [review]
While trying to delete a relation between a Prerequisite and another component, the name of Prerequisite get displayed as empty string in the popup modal.

**Review by iyannsch — CHANGES_REQUESTED** [review]
When deleting a competency, the arrows that lead to that non-existing object stay visible until the site is reloaded

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
Tested on Ts3 during testing session. Adding prerequisites to the competency overview works fine. But: circular dependencies can be added, and deleting a competency leaves the arrow behind

**Review by egekurt123 — DISMISSED** [review]
The Problem mentioned above does not exist anymore

**Review by iyannsch — CHANGES_REQUESTED** [review]
I'm still able to create circular dependencies (even without `relates` but with `matches` and `extends`) :) 
image

**Review by pzdr7 — DISMISSED** [review]
Reapprove

**Review by egekurt123 — APPROVED** [review]
Reapprove

**Review by sarpsahinalp — APPROVED** [review]
Code

**Review by az108 — APPROVED** [review]
Tested on TS6, works as expected and Code LGTM 👍

**Review by MaximilianAnzinger — DISMISSED** [review]
Maintainer approved

**Review by iyannsch — DISMISSED** [review]

**Review by egekurt123 — DISMISSED** [review]
Works fine after solving merge conflicts

**Review by iyannsch — DISMISSED** [review]
reapprove

**Review by pzdr7 — DISMISSED** [review]
Reapproving the code 👍

**Review by az108 — DISMISSED** [review]
Reapprove after merge conflict, still works 👍

**Review by az108 — APPROVED** [review]
reapprove after merge conflict

**Review by iyannsch — APPROVED** [review]
reapprove after merge conflict

**Review by sarpsahinalp — APPROVED** [review]
Reapprove after merge conflicts

**Review by SimonEntholzer — APPROVED** [review]
Re-approve after resolved merge conflicts

**Pull request merged**
